### PR TITLE
[mlir][sparse] partially support lowering sparse coiteration loops to scf.while/for.

### PR DESCRIPTION
**Pull request by PeimingLiu**
Stacked PRs:
 * #105567
 * #105566
 * __->__#105565

--- --- ---

### CODE REVIEW

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h [hunk 1]
```diff
@@ -96,24 +96,32 @@ class I64BitSet {
...
+    I64BitSet tmp = *this;
+    tmp |= p;
+    return tmp == p;
+  }
+
   // Needed by `llvm::const_set_bits_iterator_impl`.
   int find_first() const { return min(); }
   int find_next(unsigned prev) const {
-    if (prev >= max())
+    if (prev >= max() - 1)
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h:108]: any -1 or +1 always makes me a bit nervous on boundaries
is prev > max() here okay?

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h [hunk 2]
```diff
@@ -96,24 +96,32 @@ class I64BitSet {
...
       return -1;
 
-    uint64_t b = storage >> (prev + 1);
-    if (b == 0)
-      return -1;
+    uint64_t b = storage >> (prev + 1ULL);
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h:111]: avoid portability headaches with just static_cast<int64_t>(1) ?

also below?

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseIterationToScf.cpp [hunk 3]
```diff
@@ -49,6 +50,144 @@ convertIteratorType(IteratorType itTp, SmallVectorImpl<Type> &fields) {
   return success();
 }
 
+static ValueRange
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseIterationToScf.cpp:53]: It feels like this code can still be broken up into smaller utility methods to improve readabilty (probably later okay), but it is really cool to have all the loop logic generation in one file now!!

Comment (PeimingLiu) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseIterationToScf.cpp:53]: SG, will do it later!

**Review by aartbik — APPROVED** [review]

**Pull request merged**